File: tests/omp/async_target4.cpp
extern void init(  float*, int);
extern void output(float*, int);

void vec_mult(int N)
{
   int i;
   float p[N], v1[N], v2[N];
   
   {
      {
         init(v1, N);

         init(v2, N);
   
         
         for (int i=0; i<N; i++)
            p[i] = v1[i] * v2[i];
   
         output(p, N);
      }
   }
}

File: tests/omp/async_target4.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->

  !! Omp.parallel [Num_threads "2"] [tBefore; cVarDef "p"];
  !! Omp.single [] [tBefore; cSeq ~args_pred:(Target.target_list_one_st [cCall "init"]) ()];
  !! Omp.task [Depend (Out ["v1"])] [tBefore; occIndex ~nb:2 0; cCall "init"];
  !! Omp.task [Depend (Out ["v2"])] [tBefore; occIndex ~nb:2 1; cCall "init"];
  !! Omp.target [Nowait; Depend (In ["v1";"v2"]); Depend (Out ["p"]);Map_c (To, ["v1";"v2"]); Map_c (From, ["p"])] [tBefore; cFor "i"];
  !! Omp.parallel_for [Private ["i"]] [tBefore; cFor "i"];
  !! Omp.task [Depend (In ["p"])] [tBefore; cCall "output"];
)
